crates/bevy_render/src/batching/mod.rs

```diff
@@ -57,6 +68,108 @@ impl<T: PartialEq> BatchMeta<T> {
+where
+    BD: GpuArrayBufferable + Sync + Send + 'static,
+    BDI: Pod,
+{
+    /// The single buffer containing instances, used when GPU uniform building
+    /// isn't available.
+    CpuBuilt(GpuArrayBuffer<BD>),
+
+    /// The buffers containing per-instance data used when GPU uniform building
+    /// is in use.
+    GpuBuilt {
+        /// A storage area for the buffer data that the GPU compute shader is
+        /// expected to write to.
+        ///
+        /// There will be one entry for each index.
+        data_buffer: UninitBufferVec<BD>,
+
+        /// The index of the buffer data in the current input buffer that
+        /// corresponds to each instance.
+        ///
+        /// This is keyed off each view. Each view has a separate buffer.
+        work_item_buffers: EntityHashMap<Entity, BufferVec<PreprocessWorkItem>>,
+
+        /// The uniform data inputs for the current frame.
+        ///
+        /// These are uploaded during the extraction phase.
+        current_input_buffer: BufferVec<BDI>,
+
+        /// The uniform data inputs for the previous frame.
+        ///
+        /// The indices don't generally line up between `current_input_buffer`
+        /// and `previous_input_buffer`, because, among other reasons, entities
+        /// can spawn or despawn between frames. Instead, each current buffer
+        /// data input uniform is expected to contain the index of the
+        /// corresponding buffer data input uniform in this list.
+        previous_input_buffer: BufferVec<BDI>,
+    },
+}
+
+/// One invocation of the preprocessing shader: i.e. one mesh instance in a
+/// view.
+#[derive(Clone, Copy, Pod, Zeroable, ShaderType)]
+#[repr(C)]
+pub struct PreprocessWorkItem {
+    /// The index of the batch input data in the input buffer that the shader
+    /// reads from.
+    pub input_index: u32,
+    /// The index of the `MeshUniform` in the output buffer that we write to.
+    pub output_index: u32,
+}
+
+impl<BD, BDI> BatchedInstanceBuffers<BD, BDI>
+where
+    BD: GpuArrayBufferable + Sync + Send + 'static,
+    BDI: Pod,
+{
+    /// Creates new buffers.
+    pub fn new(render_device: &RenderDevice, use_gpu_uniform_builder: bool) -> Self {
+        if !use_gpu_uniform_builder {
+            return BatchedInstanceBuffers::CpuBuilt(GpuArrayBuffer::new(render_device));
+        }
+
+        BatchedInstanceBuffers::GpuBuilt {
+            data_buffer: UninitBufferVec::new(BufferUsages::STORAGE),
+            work_item_buffers: EntityHashMap::default(),
+            current_input_buffer: BufferVec::new(BufferUsages::STORAGE),
+            previous_input_buffer: BufferVec::new(BufferUsages::STORAGE),
+        }
+    }
+
+    /// Returns the binding of the uniform buffer that contains the per-instance
+    /// data.
+    ///
+    /// If we're in the GPU uniform building mode, this buffer needs to be
+    /// filled in via a compute shader.
+    pub fn uniform_binding(&self) -> Option<BindingResource> {
```

Comment:
```suggestion
    pub fn binding(&self) -> Option<BindingResource> {
```

Reply:
There are multiple buffers in there, so I'll call it `instance_data_binding` for clarity.

Reply:
Works for me